### PR TITLE
Devel g5 fix

### PR DESCRIPTION
**Pull request by dvstans**
This PR changes the handling of the Globus 5 base path for repository endpoints. The base (and repo) paths are now set in the repo configuration in the DB and the repo server expects full paths to be sent to it. This fix has been tested on Globus 5; however, it will likely break Globus 4 installations. (No Globus 4 test server is available.) Note that Globus 4 is being deprecated soon.

### CODE REVIEW

File: core/database/api/repo_router.js [hunk 1]
```diff
@@ -92,48 +143,26 @@ router.post('/create', function (req, res) {
...
                     throw g_lib.ERR_PERM_DENIED;
 
-                var obj = {
-                    capacity: req.body.capacity,
-                    pub_key: req.body.pub_key,
-                    address: req.body.address,
-                    endpoint: req.body.endpoint,
-                    path: req.body.path
-                };
+                var obj = parseRepoDoc( req.body );
```
Comment (JoshuaSBrown) [inline core/database/api/repo_router.js:146]: I love changes like this, this makes things so much more testable.

File: core/database/api/repo_router.js [hunk 2]
```diff
@@ -92,48 +143,26 @@ router.post('/create', function (req, res) {
...
-                        obj.path += "/";
-                }
 
                 var repo = g_db.repo.save( obj, { returnNew: true });
 
                 for ( var i in req.body.admins ) {
-                    if ( !g_db._exists( req.body.admins[i] ))
+                    if ( !g_db._exists( req.body.admins[i] )){
                         throw [g_lib.ERR_NOT_FOUND,"User, "+req.body.admins[i]+", not found"];
+                    }
```
Comment (JoshuaSBrown) [inline core/database/api/repo_router.js:155]: This is an improvement, I have hit several errors when I have accidentally deleted white space in front of code that had no curly braces.

File: core/database/api/repo_router.js [hunk 3]
```diff
@@ -92,48 +143,26 @@ router.post('/create', function (req, res) {
...
-                var obj = {
-                    capacity: req.body.capacity,
-                    pub_key: req.body.pub_key,
-                    address: req.body.address,
-                    endpoint: req.body.endpoint,
-                    path: req.body.path
-                };
+                var obj = parseRepoDoc( req.body );
 
                 g_lib.procInputParam( req.body, "id", false, obj );
```
[the remark this answers is not in the document]
Comment (dvstans) [reply core/database/api/repo_router.js:148]: The ID field is only present in the create method not the update. Could have been checked in the parse function either way.

File: core/database/api/repo_router.js (hunk 2 above)
Comment (dvstans) [reply core/database/api/repo_router.js:155]: Yeah, I'm trying to fix these when I find them / think about it.

File: core/database/api/repo_router.js [hunk 4]
```diff
@@ -146,18 +175,16 @@ router.post('/create', function (req, res) {
     id: joi.string().required(),
     title: joi.string().required(),
     desc: joi.string().optional(),
-    domain: joi.string().optional(),
-    capacity: joi.number().integer().min(0).required(),
+    capacity: joi.number().integer().min(1).required(),
```
Comment (JoshuaSBrown) [inline core/database/api/repo_router.js:178]: What is the reason for this change? Are you just wanting to ensure that when someone creates a repo they actually assign something to it?

File: common/proto/SDMS_Auth.proto [hunk 5]
```diff
@@ -907,14 +907,13 @@ message RepoCreateRequest
     required string             id          = 1; // ID of repo
     required string             title       = 2; // Title
     optional string             desc        = 3; // Description
-    optional string             domain      = 5; // RESERVED
-    required string             path        = 6; // Path to storage directories
-    optional string             exp_path    = 7; // RESERVED
```
Comment (JoshuaSBrown) [inline common/proto/SDMS_Auth.proto:912]: I'm not following what "exp_path" is actually for from the Globus 4?

File: core/database/api/repo_router.js (hunk 4 above)
Comment (dvstans) [reply core/database/api/repo_router.js:178]: are you referring to the capacity? if so, a repo must have a non-zero capacity to make sense. The original min(0) was a mistake.

File: common/proto/SDMS_Auth.proto (hunk 5 above)
Comment (dvstans) [reply common/proto/SDMS_Auth.proto:912]: It was intended to be used in conjunction with the FUSE driver to tell end-users what the local path to the controlled data in the repo was (so they could open it directly). None of this will work going forward if we want to support non-posix storage options.

File: core/database/api/repo_router.js (hunk 4 above)
Comment (JoshuaSBrown) [reply core/database/api/repo_router.js:178]: Makes sense.

File: repository/server/RequestWorker.cpp [hunk 6]
```diff
@@ -279,10 +281,11 @@ RequestWorker::procDataGetSizeRequest()
         }
         else
         {
-            data_sz->set_size( 0 );
```
Comment (JoshuaSBrown) [inline repository/server/RequestWorker.cpp:282]: Why did you remove this line?

Comment (dvstans) [reply repository/server/RequestWorker.cpp:282]: Odd, this is not showing the latest code changes, but it was replaced with an exception due to not finding the specified file path. The prior code would allow a transfer to succeed if there was a repo issue.

File: web/static/dlg_repo_edit.js [hunk 7]
```diff
@@ -188,7 +184,7 @@ export function show( a_repo_id, a_cb ){
                     node.data.alloc.totalSz = data.totalSz;
                     updateAllocTitle( node );
 
-                    var msg =
+                    var msg =a_repo_id
```
Comment (JoshuaSBrown) [inline web/static/dlg_repo_edit.js:187]: So this looks like it is only sending the repo_id back is that the intention or is my javascript off?

Comment (dvstans) [reply web/static/dlg_repo_edit.js:187]: No, that's a weird typo. Not sure how that got there but needs to be fixed (there's a msg body below this not shown).

File: web/static/dlg_repo_edit.js [hunk 8]
```diff
@@ -118,15 +114,15 @@ export function show( a_repo_id, a_cb ){
...
     if ( a_repo_id ){
-        $("#title",frame).on('input', function(){ repoInputChanged(1); });
-        $("#desc",frame).on('input', function(){ repoInputChanged(2); });
-        $("#domain",frame).on('input', function(){ repoInputChanged(4); });
-        $("#capacity",frame).on('input', function(){ repoInputChanged(8); });
-        $("#path",frame).on('input', function(){ repoInputChanged(0x10); });
-        $("#exp_path",frame).on('input', function(){ repoInputChanged(0x20); });
-        $("#pub_key",frame).on('input', function(){ repoInputChanged(0x40); });
-        $("#addr",frame).on('input', function(){ repoInputChanged(0x80); });
-        $("#ep_id",frame).on('input', function(){ repoInputChanged(0x100); });
```
Comment (JoshuaSBrown) [inline web/static/dlg_repo_edit.js:129]: This is definately cleaner and more readable getting rid of the magick numbers, though I'm not sure I fully grasp the implications of the change, I'm guessing you were using a bitmask to specify what parts needed to be updated and are now instead just doing a blanket update of the form?

File: repository/server/RequestWorker.cpp (hunk 6 above)
Comment (JoshuaSBrown) [reply repository/server/RequestWorker.cpp:282]: Tyler and I noticed the other day that GitHub diffs were not showing the most recent changes perhaps this is the same problem.

File: web/static/dlg_repo_edit.js (hunk 8 above)
Comment (dvstans) [reply web/static/dlg_repo_edit.js:129]: Yep, no need to support partial updates.

File: web/static/dlg_repo_edit.js [hunk 9]
```diff
@@ -274,97 +270,60 @@ export function show( a_repo_id, a_cb ){
...
-                    if ( cap == null ){
-                        dialogs.dlgAlert("Data Entry Error","Invalid repo capacity value." );
-                        return;
-                    }
-                    if ( cap != repo.capacity )
-                        obj.capacity = cap;
+                    obj.id = a_repo_id;
+                } else {
+                    obj.id = $("#id",frame).val();
+                }
```
Comment (JoshuaSBrown) [inline web/static/dlg_repo_edit.js:291]: This is a lot clearer.

**Review by JoshuaSBrown — APPROVED** [review]
There is only one thing that still needs to be updated, there is a helper script that tells admins how to fill in the repo form in /scripts/globus it probably needs to be updated to reflect the new fields, but... its not really a blocking problem.
